### PR TITLE
Loan transaction reversal data extension

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

Upon reversing a loan transaction, 2 new data fields need to be added. One is the ability to provide an external ID for the reversal and one for storing when the reversal happened.

- The external ID for the reversal is important because other systems might track reversal transactions separately. 

- The reversal date is also important since the Loan transactions have a submittedOnDate, therefore it makes sense to store when a particular transaction has been reversed. It is set with the Business date value when the Loan transaction is undo

Undo a Loan Transaction
Screen Shot 2022-08-08 at 19 59 02

Get the Loan Transaction reversed
Screen Shot 2022-08-08 at 20 17 20

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by adamsaghy** [comment]
@josehernandezfintecheandomx Can you please take a look on the conflicts?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransaction.java [hunk 1]
```diff
@@ -106,6 +106,12 @@ public class LoanTransaction extends AbstractAuditableWithUTCDateTimeCustom {
     @Column(name = "external_id", length = 100, nullable = true, unique = true)
     private String externalId;
 
+    @Column(name = "reversal_external_id", length = 100, nullable = true, unique = true)
+    private String reversalExternalId;
+
+    @Column(name = "reversed_on_date", nullable = false)
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransaction.java]: @josehernandezfintecheandomx It should be nullable!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java [hunk 2]
```diff
@@ -213,6 +207,11 @@ public void validateTransaction(final String json) {
         final String note = this.fromApiJsonHelper.extractStringNamed("note", element);
         baseDataValidator.reset().parameter("note").value(note).notExceedingLengthOf(1000);
 
+        final String reversalExternalId = this.fromApiJsonHelper.extractStringNamed(LoanApiConstants.REVERSAL_EXTERNAL_ID_PARAMNAME,
+                element);
+        baseDataValidator.reset().parameter(LoanApiConstants.REVERSAL_EXTERNAL_ID_PARAMNAME).ignoreIfNull().value(reversalExternalId)
+                .notExceedingLengthOf(1000);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java]: @josehernandezfintecheandomx Based on the liquibase script, the size of this string cannot be longer than 100

**Review by adamsaghy — CHANGES_REQUESTED** [review]
@josehernandezfintecheandomx Please check my comments and also please add an integration test which checks whether the reversal date and provided external id is stored and fetched appropriately!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransaction.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransaction.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java]: Done!

**Comment by adamsaghy** [comment]
@josehernandezfintecheandomx Please take a look on the conflicts!

**Comment by josehernandezfintecheandomx** [comment]
> @josehernandezfintecheandomx Please take a look on the conflicts!

Done @adamsaghy

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/TextUtils.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ */
+package org.apache.fineract.infrastructure.core.service;
+
+import org.apache.commons.lang3.StringUtils;
+
+public final class TextUtils {
+
+    private TextUtils() {}
+
+    public static boolean is(final String commandParam, final String commandValue) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/TextUtils.java]: I see you extracted this from the other class.
I think the TextUtils is very generic, it rather should be something like CommandParameterUtil or something with a more descriptive name.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransaction.java [hunk 4]
```diff
@@ -837,6 +849,10 @@ public LocalDate getSubmittedOnDate() {
         return submittedOnDate;
     }
 
+    public void setReversalExternalId(String reversalExternalId) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransaction.java]: I don't think we wanna allow to override the reversal external ID once its set, do we?
I feel the reverse method should be more than enough to provide the reversal external ID.

**Review by galovics — CHANGES_REQUESTED** [review]
Almost done, good job. Please check the remaining 2 comments and we can merge.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/TextUtils.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/TextUtils.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransaction.java (hunk 4 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransaction.java]: Done! Removed

**Review by galovics — APPROVED** [review]
LGTM

**Comment by galovics** [comment]
@josehernandezfintecheandomx please resolve the conflicts.

**Comment by josehernandezfintecheandomx** [comment]
> @josehernandezfintecheandomx please resolve the conflicts.

The conflicts are solved

**Pull request merged**